### ISSUE TITLE
numref link in PDF jumps to the wrong location

### ISSUE THREAD

**Issue by andreacassioli** [sphinx-doc/sphinx#2405]
Hi, 
I use numref to generate links to code listing. For instance I have

```
 :numref:`toolbox.code_sdo1` demonstrates how to solve this problem.


   .. literalinclude:: stripex/sdo1.m
      :name: toolbox.code_sdo1
      :caption: Code implementing problem :eq:`tutorial_sdo1`.

```

The listing is quite long and occupies 2 and half pdf pages. The link I get seems correct, but if I click I jump to the end of the listing instead the beginning, where even the caption is. This is not right!

**Comment by jfbu** [comment]
Thus, this is LaTeX builder issue. Indeed, generated tex source has

``` latex
 \end{Verbatim}
 \phantomsection\label{index:toolbox-code-sdo1}
```

_after_  the Verbatim environment. It should be moved earlier.

Brief testing shows that if the produced latex file contains rather

``` latex
\phantomsection\label{index:toolbox-code-sdo1}
\begin{Verbatim}[commandchars=\\\{\}]
```

then the link works fine. Also when the caption just barely made it to start of a page, the link goes to right page.

**Comment by jfbu** [comment]
@andreacassioli could you try the patch  to `latex.py` at https://github.com/jfbu/sphinx/commit/083d8f8e0e0f2e149ebf0723b584445764f24c69 on a real life project. I am not familiar enough with Sphinx to evaluate if it can impact something else. The test suite (on master branch at 55a991e) ran ok, apart from an error which is completely unrelated:

```
  File "[...]/sphinx/tests/test_config.py", line 13, in <module>
    from mock import patch
ImportError: No module named 'mock'
```

ah ok, this is because I ran test suite on master at 55a991e not on stable which has a recent commit 0c33d7df regarding mock module.

**Comment by andreacassioli** [comment]
@jfbu I have tried on a dummy project made from scratch and it seems to work well. Good job! It would be nie to have it as bug fixing in 1.4.1.

**Comment by jfbu** [comment]
@andreacassioli Thanks for testing; good to know it seems to work. I can make a PR but I feel my modifications first need some review by people who understand globally `latex.py`. 

In the meantime I have put the patch on top of stable (modified minor things related to location of `\n` in the strings) there https://github.com/jfbu/sphinx/commit/6123ea49e53b57e200138744dc618a797633d4d6

**Comment by jfbu** [comment]
@andreacassioli  hmm, examining more closely the situation from the `index.rst` at top of thread, I see that although the hyperlink works, the LaTeX situation is not ideal: the tex file contains:

```
\hyperref[index:toolbox-code-sdo1]{Listing \ref{index:toolbox-code-sdo1}} demonstrates how to solve this problem.
```

But the `\ref{index:toolbox-code-sdo1}` does not work, because this is a LaTeX mechanism which needs a  `\refstepcounter{foo}` somewhere. And there is none. One sees `Listing<space>` and not `Listing 1` for the hyperref link.

**Comment by jfbu** [comment]
@andreacassioli Sorry to ask again, but is it possible you test https://github.com/jfbu/sphinx/tree/jumptorightplace-onstable on a real project? (this is a branch based on stable commit d35ccb9b4c3b15dbfd81e8349f0a1464f2e423d8, there are currently two commits and the tip is at https://github.com/jfbu/sphinx/commit/a1647b6bb9f89141a9ba5b3579cc1bce8d4b18c2, the first commit is equivalent to the one you already tested). I am not sure but I believe with previous code you would not see `Listing <n>` at the location of the hyperlinks but only `Listing` without numeric reference. The last commit I did attempts to fix this, fingers crossed that it does not introduce incompatibilities. I am not familiarized enough with Sphinx source code to be sure.

**Comment by jfbu** [comment]
I added a commit to https://github.com/jfbu/sphinx/tree/jumptorightplace-onstable to take out a test I did before a `context.append(ids)`. The context of my hesitation is that I am not sure  whether the move of `context.pop()`  to `visit_caption`  could cause a problem if there is no caption, is it possible that the code doing `context.append(ids)` is executed also in situations without caption, hence what is appended is not properly removed ? this is a question to people who know Sphinx's `latex.py`.

**Comment by andreacassioli** [comment]
Hi, you are right. Your fix works well if a caption is available, otherwise
- the code has no caption, not even `List 1` or the like
- there is no link at all, just a verbatim `list`

The Tex code I see is 

```
See \code{list}.

\begin{Verbatim}[commandchars=\\\{\}]
% the code
\end{Verbatim}
```

I think this is the way `numref` works, in the sense that i my understanding only if a title is availble a link is generated. I think the fix should be in the numref role itself.

**Comment by jfbu** [comment]
Agreed about what you observe in absence of a caption. But just to be clear this is already the situation on stable branch of Sphinx, it is not related to my fix of the original issue in this thread.

**Comment by jfbu** [comment]
Also, when testing a `numref:`foo`` in the situation with no caption, I do get a Sphinx warning on `make latex`. 

```
processing testnumref.tex... index 
resolving references...
/xxxxxxxxxx/2405numref/index.rst:1: WARNING: undefined label: foo
writing... done
copying TeX support files...
done
build succeeded, 1 warning.
```

Tested on stable branch. My fix in https://github.com/jfbu/sphinx/tree/jumptorightplace-onstable has no impact on this, it is about making sure, when `:caption:` is made use of, that `numref` will generate not only correct link but also correct numbering of literal block for the link text in PDF.

**Comment by jfbu** [comment]
Both this and #2414 are addressed by PR #2415

**Comment by andreacassioli** [comment]
Hi @jfbu ,
thanks for your effort. I hope to see the patches son in stable!

**Comment by tk0miya** [comment]
Fixed by #2415 :-)
